src/codegen/compile.ml

```diff
@@ -7194,7 +7194,8 @@ module MakeSerialization (Strm : Stream) = struct
                 continue
             )
             ( sort_by_hash vs )
-            ( coercion_failed "IDL error: unexpected variant tag" )
+            ( skip get_arg_typ ^^ (* HERE *)
+              coercion_failed "IDL error: unexpected variant tag" )
```

Comment:
Well spotted. The fix looks reasonable.

Reply:
According to git blame, I made that mistake in https://github.com/dfinity/motoko/commit/5388dc3af52b16c2b34440f64f8fd221cda060d5 and it hasn't surfaced in three years. Oh dear.

Reply:
Great, thanks for checking! I'll try to produce a smaller test case

Reply:
Ideally include some tests in the official Candid test suite, so that other implementations don't make that mistake (although I believe all others first parse and then coerce, and thus are less prone to this bug).

An alternative implementation would be to return quickly, and when backtracking in `opt` reset the data and type but pointers and skip the whole thing in one go. Might be more reliable and simpler to implement, especially for records.

Reply:
Yeah, I was wondering about that but don't trust myself to rewrite this code. Is the current invariant that the recursive call needs to finish parsing the entire value, leaving the buffer and data pointers just past the end of the value?

Reply:
Yes, and I think there is a comment to that end somewhere.